### PR TITLE
#168. Migrate ListView to RecyclerView in MainActivity.

### PR DESCRIPTION
**Pull request by bogdan-evtushenko**

### CODE REVIEW

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/ui/presenter/ShortSummaryPresenter.java [hunk 1]
```diff
@@ -102,17 +133,37 @@ private String formatPeriod(Period period) {
     }
 
     public static class ViewHolder {
-        @BindView(R.id.tv_period)
+
+        @BindView(R.id.tvPeriod)
         TextView tvPeriod;
-        @BindView(R.id.tv_total_income)
+
+        public TextView getTvPeriod() {
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/blogspot/e_kanivets/moneytracker/ui/presenter/ShortSummaryPresenter.java]: No need for getters here. Just make fields public.

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/ui/presenter/ShortSummaryPresenter.java [hunk 2]
```diff
@@ -45,40 +49,67 @@ public ShortSummaryPresenter(Context context) {
...
     public void update(IRecordReport report, String currency, List<String> ratesNeeded) {
-        ViewHolder viewHolder = (ViewHolder) view.getTag();
-
-        if (report == null) {
-            viewHolder.tvTotalIncome.setText("");
-            viewHolder.tvTotalExpense.setText("");
-
-            viewHolder.tvTotal.setTextColor(red);
-            viewHolder.tvTotal.setText(createRatesNeededList(currency, ratesNeeded));
+        if (view.getTag() instanceof RecordAdapter.HeaderViewHolder) {
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/blogspot/e_kanivets/moneytracker/ui/presenter/ShortSummaryPresenter.java]: Hm, you can extract an interface with `getTvTotalName(), ...` and then have both ViewHolders implementing them. In such a way you won't have duplication (and class check) here.

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/activity/record/MainActivity.kt [hunk 3]
```diff
@@ -125,7 +122,7 @@ class MainActivity : BaseDrawerActivity() {
                 REQUEST_ACTION_RECORD -> update()
 
                 REQUEST_BACKUP -> {
-                    appComponent.inject(this@MainActivity)
```
[the remark this answers is not in the document]
Comment (bogdan-evtushenko) [reply app/src/main/java/com/blogspot/e_kanivets/moneytracker/activity/record/MainActivity.kt]: What do you mean?)

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+        whiteRed = ContextCompat.getColor(context, R.color.white_red)
+        whiteGreen = ContextCompat.getColor(context, R.color.white_green)
+        red = ContextCompat.getColor(context, R.color.red)
+        green = ContextCompat.getColor(context, R.color.green)
+
+        summaryPresenter = ShortSummaryPresenter(context)
+
+        if (isHeaderViewNeeded) {
+            headerViewHolder = HeaderViewHolder(LayoutInflater.from(context).inflate(R.layout.view_summary_records, null), itemClickListener)
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordAdapter.kt]: I think, you can move this code to `onCreateViewHolder`.

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    fun setRecords(recordsList: List<Record>, report: IRecordReport?, currency: String, ratesNeeded: List<String>) {
+        records = recordsList
+        summaryPresenter.update(report, currency, ratesNeeded)
+        notifyDataSetChanged()
+    }
+
+    fun setOnItemClickListener(itemClickListener: OnItemClickListener) {
+        this.itemClickListener = itemClickListener
+        if (isHeaderViewNeeded) {
+            headerViewHolder = HeaderViewHolder(LayoutInflater.from(context).inflate(R.layout.view_summary_records, null), itemClickListener)
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordAdapter.kt]: It's not clear why we need this 🤔

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            tvPrice = view.tvPrice
+            tvTitle = view.tvTitle
+            tvCategory = view.tvCategory
+            tvCurrency = view.tvCurrency
+
+            view.setOnClickListener {
+                itemClickListener?.onItemClick(adapterPosition)
+            }
+        }
+
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordAdapter.kt]: Empty line 😉

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    class HeaderViewHolder : RecyclerView.ViewHolder, ShortSummaryPresenter.SummaryViewInterface {
+
+        private var tvPeriod: TextView
+        private var tvTotalIncome: TextView
+        private var tvTotalExpense: TextView
+        private var tvTotal: TextView
+
+        override fun getTvPeriod(): TextView {
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordAdapter.kt]: Can be as simple as `override fun getTvPeriod() = tvPeriod`

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        fun onItemClick(position: Int)
+    }
+
+    companion object {
+
+        private const val TYPE_HEADER = 0
+        private const val TYPE_ITEM = 1
+
+    }
+
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordAdapter.kt]: Empty lines 🏭

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/ui/presenter/ShortSummaryPresenter.java [hunk 9]
```diff
@@ -45,40 +46,42 @@ public ShortSummaryPresenter(Context context) {
...
     }
 
-    public View create(boolean shortSummary) {
+    public View create(boolean shortSummary, RecordAdapter.HeaderViewHolder mainViewHolder) {
         view = layoutInflater.inflate(R.layout.view_summary_records, null);
         view.findViewById(R.id.iv_more).setVisibility(shortSummary ? View.VISIBLE : View.INVISIBLE);
 
-        ViewHolder viewHolder = new ViewHolder(view);
-        view.setTag(viewHolder);
+        if (mainViewHolder == null)
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/blogspot/e_kanivets/moneytracker/ui/presenter/ShortSummaryPresenter.java]: You can do `view.setTag(mainViewHolder == null ? new ViewHolder(view) : mainViewHolder);`

**Comment by bogdan-evtushenko** [comment]
@yevhenii-kanivets review please

**Comment by yevhenii-kanivets** [comment]
@bogdan-evtushenko following bug has been found while testing:

`Short Summary` at Records Screen isn't updated when records are added on the first launch. Any operations actually don't affect summary.

**Pull request merged**